File: solutions/179_largest_number.cpp
#include "aux.cpp"
/**
Given a list of non negative integers, arrange them such that they form the
largest number.
Example 1:
Input: [10,2]
Output: "210"
Example 2:
Input: [3,30,34,5,9]
Output: "9534330"
Note: The result may be very large, so you need to return a string instead of an
integer.

 https://leetcode.com/problems/largest-number/description/
 **/
static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();
class Solution {
public:
  string largestNumber(vector<int> &nums) {
    if (nums.empty())
      return "";
    sort(nums.begin(), nums.end(), [](const int &a, const int &b) {
      return to_string(b) + to_string(a) < to_string(a) + to_string(b);
    });
    string res = "";
    for (int i = 0; i < nums.size(); i++)
      res.append(to_string(nums[i]));
    return res[0] == '0' ? "0" : res;
  }
};
int main() {
  Solution s;
  vector<int> nums = {31, 309, 47, 5, 9};
  nums = {0, 0, 1};
  say(s.largestNumber(nums));
}

File: solutions/179_largest_number.rb
# Given a list of non negative integers, arrange them such that they form the largest number.
# Example 1:
# Input: [10,2]
# Output: "210"
# Example 2:
# Input: [3,30,34,5,9]
# Output: "9534330"
# Note: The result may be very large, so you need to return a string instead of an integer.
#
#  https://leetcode.com/problems/largest-number/description/
require './aux.rb'

# @param {Integer[]} nums
# @return {String}

# a should be 'larger'
def comp(a, b)
  0.upto([a.size, b.size].min - 1).each do |i|
    return 1 if a[i] > b[i]
    return -1 if a[i] < b[i]
  end
  return 0 if a.size == b.size
  a.size > b.size ? comp(a[b.size..-1], b) : comp(a, b[a.size..-1])
end

def largest_number_2(nums)
  nums.map(&:to_s).sort { |a, b| -comp(a, b) }.join.to_i.to_s
end

def largest_number(nums)
  nums.sort { |a, b| b.to_s + a.to_s <=> a.to_s + b.to_s }.join.to_i.to_s
end

nums = [3, 30, 34, 5, 9]
# nums = [10, 2, 100]
# nums = [0, 0]

p largest_number(nums)
